### ISSUE TITLE
new_ret_no_self complains about new() -> impl Future<Self> if new() takes parameters

### ISSUE THREAD

**Issue by Arnavion** [rust-lang-nursery/rust-clippy#3220]
`0.0.212 (2018-09-17 125907a)`

[Playground](https://play.rust-lang.org/?gist=b9ba6a68babfe77aeb8ef34058a9b732&version=nightly&mode=debug&edition=2015)

```rust
#![feature(tool_lints)]

#![deny(clippy::new_ret_no_self)]

extern crate futures;
use futures::{Future, IntoFuture};

struct S;

impl S {
    pub fn new(_: String) -> impl Future<Item = Self, Error = ()>
    {
        Ok(S).into_future()
    }
}

fn main() {}
```

```
error: methods called `new` usually return `Self`
  --> src/main.rs:11:5
   |
11 | /     pub fn new(_: String) -> impl Future<Item = Self, Error = ()>
12 | |     {
13 | |         Ok(S).into_future()
14 | |     }
   | |_____^
   |
note: lint level defined here
  --> src/main.rs:3:9
   |
3  | #![deny(clippy::new_ret_no_self)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^
   = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#new_ret_no_self
```

If `S::new()` does not have the `String` parameter, the lint does not complain.

**Comment by Arnavion** [comment]
"The compiler unexpectedly panicked" happens if the compiler API panics. clippy uses the compiler API. So yes, I would start by assuming it's from your change.

**Comment by mati865** [comment]
@JoshMcguigan run the test with `RUST_BACKTRACE=1` and check the trace. If you need more help you can create WIP PR.

**Comment by JoshMcguigan** [comment]
I opened a PR (#3253) which resolves this issue as well as another issue with this lint, but I'm not sure I did it in the best possible way. I'd be very open to any feedback.
